### PR TITLE
[Unity][OP] Add an operator for fused multi head attention

### PR DESCRIPTION
**Pull request by cyx-6**
This PR introduces the new relax operator `R.nn.attention` for fused multi head attention, and the support of fused multi head attention to relax cutlass BYOC. The input of the operator are query, key and value tensor, with `BSNH` layout, namely `[batch size, sequence length, number of heads, dimension of heads]`. And the output shares the same layout with all input tensor.

### CODE REVIEW

**Comment by tqchen** [comment]
thanks @cyx-6 , can we also add a note about existing attention operator layout in libraries(HF and PT), that would help us to write down the rationale of the operator

File: tests/python/relax/test_codegen_cutlass.py [hunk 1]
```diff
@@ -438,5 +439,79 @@ def test_matmul_transposed_bias_gelu_offload(matmul_x, matmul_y, matmul_bias):
...
+                result = R.emit(R.nn.attention(q, k, v))
+                R.output(result)
+
+            R.func_ret_value(frame.output_vars[0])
+
+    func = builder.get()
+    return tvm.IRModule({"main": func})
+
+
+def get_numpy_attention_ref(q, k, v):
```
Comment (tqchen) [inline tests/python/relax/test_codegen_cutlass.py:495]: use tvm.contrib.memoize so we don't need to run it multiple times

File: src/relax/op/nn/attention.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+#include <tvm/relax/attrs/nn.h>
+
+#include "../op_common.h"
+
+namespace tvm {
+namespace relax {
+
+/*! \brief fused multi head attention */
+Expr attention(Expr query, Expr key, Expr value, DataType out_dtype);
```
Comment (masahi) [inline src/relax/op/nn/attention.h:36]: what about bias?

**Review by masahi — COMMENTED** [review]
Please add a test case for the Relax op.

File: python/tvm/contrib/cutlass/attention_operation.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+  CHECK(${arg0}->ndim == 4); // B, S, N, H
+  CHECK(${arg1}->ndim == 4); // B, S', N, H
+  CHECK(${arg2}->ndim == 4); // B, S', N, H'
+  CHECK(out0->ndim == 4); // B, S, N, H'
+
+  using Attention =
+      AttentionKernel<T,
+                      /*ArchTag=*/${arch},
+                      /*is_aligned=*/true,
```
Comment (vinx13) [inline python/tvm/contrib/cutlass/attention_operation.py:37]: we also need to dispatch `is_aligned` based on input shape

File: python/tvm/contrib/cutlass/attention_operation.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  p.q_strideM = p.q_strideH * p.num_heads; // H * N
+  p.k_strideM = p.k_strideH * p.num_heads; // H * N
+  p.v_strideM = p.v_strideH * p.num_heads; // H' * N
+  p.o_strideM = p.head_dim_value * p.num_heads; // H' * N
+
+  // stride for B
+  p.q_strideB = p.q_strideM * p.num_queries; // H * N * S
+  p.k_strideB = p.k_strideM * p.num_keys; // H * N * S'
+  p.v_strideB = p.v_strideM * p.num_keys; // H'* N * S'
+  // p.o_strideB = p.o_strideM * p.num_queries; // H'* N * S
```
Comment (vinx13) [inline python/tvm/contrib/cutlass/attention_operation.py:78]: remove if not needed

File: python/tvm/contrib/cutlass/attention_operation.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name, unused-wildcard-import, wildcard-import
+"""Generator for CUTLASS attention kernels."""
+from .library import *
+
+
+def instantiate_attention_template(attrs, func_args):
```
Comment (sunggg) [inline python/tvm/contrib/cutlass/attention_operation.py:22]: Correct me if i'm wrong, but I heard there are various ways to implement attention. 
Is there any specific form (maybe something standard) of attention we target? 
Or can we also support a set of variants?

**Review by sunggg — COMMENTED** [review]
Thank you for exciting contribution, @cyx-6! 
Overall, it looks good. Please see my comments below.

File: python/tvm/contrib/cutlass/attention_operation.py (hunk 5 above)
Comment (vinx13) [reply python/tvm/contrib/cutlass/attention_operation.py:22]: this only implements `softmax(Q*K.T) * V` in BSNH layout. We can support other layout, and fused QKV in the future by changing the strides

**Comment by spectrometerHBH** [comment]
Do we consider adding it to the Fx translator since PT has MHA op?
https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html

**Comment by masahi** [comment]
cc @hwu36

**Comment by hwu36** [comment]
@danthe3rd @mnicely

File: tests/python/relax/test_codegen_cutlass.py (hunk 1 above)
Comment (cyx-6) [reply tests/python/relax/test_codegen_cutlass.py]: Sure! I have refactored the unittest codes and add the memoize decorator.

File: include/tvm/relax/attrs/nn.h [hunk 6]
```diff
@@ -184,6 +184,15 @@ struct DropoutAttrs : public tvm::AttrsNode<DropoutAttrs> {
   }
 };  // struct DropoutAttrs
 
+/*! \brief Attributes used in fuse multi head attention operator */
+struct AttentionAttrs : public tvm::AttrsNode<AttentionAttrs> {
+  DataType out_dtype;
```
[the remark this answers is not in the document]
Comment (cyx-6) [reply include/tvm/relax/attrs/nn.h]: Yes. The `AttentionAttrs` is redundant indeed. I have removed them.

File: python/tvm/contrib/cutlass/attention_operation.py (hunk 4 above)
Comment (cyx-6) [reply python/tvm/contrib/cutlass/attention_operation.py]: Sure. The commented codes have been removed.

File: python/tvm/contrib/cutlass/attention_operation.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  CHECK(out0->ndim == 4); // B, S, N, H'
+
+  using Attention =
+      AttentionKernel<T,
+                      /*ArchTag=*/${arch},
+                      /*is_aligned=*/true,
+                      /*queries_per_block=*/${kQueriesPerBlock},
+                      /*keys_per_block=*/${kKeysPerBlock},
+                      /*single_value_iteration=*/${kSingleValueIteration}
+      >;
```
Comment (danthe3rd) [inline python/tvm/contrib/cutlass/attention_operation.py:41]: Note: if you don't need it, you should set `kSupportsDropout` and `kSupportsBias` to false for better performance
https://github.com/NVIDIA/cutlass/blob/f396cdd15ccf873d5c92ae73b73ed680c77f4400/examples/41_fused_multi_head_attention/kernel_forward.h#L107-L108

File: python/tvm/contrib/cutlass/attention_operation.py (hunk 5 above)
Comment (slyubomirsky) [reply python/tvm/contrib/cutlass/attention_operation.py:22]: Is that the difference between Bahdanau and Luong attention? (I don't have a ton of experience with attention mechanisms)

File: src/relax/op/nn/attention.h (hunk 2 above)
Comment (vinx13) [reply src/relax/op/nn/attention.h:36]: yeah bias can be useful in some cases. @cyx-6 can we add an optional input `bias` for attention op?

Comment (cyx-6) [reply src/relax/op/nn/attention.h:36]: yes, I am adding it

**Comment by MasterJH5574** [comment]
Just note that 3rdparty/cutlass is changed by this PR. Could you revert this change?

File: python/tvm/contrib/cutlass/attention_operation.py (hunk 5 above)
Comment (vinx13) [reply python/tvm/contrib/cutlass/attention_operation.py:22]: it's mostly different variants of mask/bias apply to Q*K.T

File: python/tvm/contrib/cutlass/gen_tensor_op.py [hunk 8]
```diff
@@ -652,4 +653,25 @@ def get_batch_stride(stride_annot, arg0_idx, arg1_idx, arg0_axis_idx, arg1_axis_
         code = instantiate_conv2d_template(attrs, func_args)
         return CodegenResult(code, headers)
 
+    elif "attention" in func_name:
+        headers.append("kernel_forward.h")
+        attrs["num_batches"] = str(int(annotations["num_batches"]))
+        attrs["num_queries"] = str(int(annotations["num_queries"]))
+        attrs["num_keys"] = str(int(annotations["num_keys"]))
+        attrs["num_heads"] = str(int(annotations["num_heads"]))
+        attrs["head_dim"] = str(int(annotations["head_dim"]))
```
Comment (masahi) [inline python/tvm/contrib/cutlass/gen_tensor_op.py:662]: Feel free to update `substitute_template(...)` to get rid off `str(int(...))` stuff. I can also do it later otherwise.

File: python/tvm/contrib/cutlass/attention_operation.py (hunk 3 above)
Comment (cyx-6) [reply python/tvm/contrib/cutlass/attention_operation.py]: the dispatch for alignment has been added

File: python/tvm/contrib/cutlass/gen_tensor_op.py (hunk 8 above)
Comment (cyx-6) [reply python/tvm/contrib/cutlass/gen_tensor_op.py]: The `substitute_template` is now updated to accept `int`, `IntImm` and `bool`. And all the usages of `str(int(...))` have been removed, including usages in conv2d and matmul.

Comment (masahi) [reply python/tvm/contrib/cutlass/gen_tensor_op.py]: Thanks!

File: python/tvm/contrib/cutlass/gen_tensor_op.py [hunk 9]
```diff
@@ -653,4 +655,34 @@ def get_batch_stride(stride_annot, arg0_idx, arg1_idx, arg0_axis_idx, arg1_axis_
...
+            attrs["kIsAligned"] = False
+        else:
+            raise NotImplementedError()
+        if h_v > 64:
+            attrs["kQueriesPerBlock"] = "32"
+            attrs["kKeysPerBlock"] = "128"
+            attrs["kSingleValueIteration"] = h_v <= 128
+        else:
+            attrs["kQueriesPerBlock"] = "64"
+            attrs["kKeysPerBlock"] = "64"
```
Comment (masahi) [inline python/tvm/contrib/cutlass/gen_tensor_op.py:681]: We can also remove the quotation now.

Comment (cyx-6) [reply python/tvm/contrib/cutlass/gen_tensor_op.py]: Oh, I missed that. Fixed already.

File: src/relax/op/nn/attention.h (hunk 2 above)
Comment (cyx-6) [reply src/relax/op/nn/attention.h]: The bias for `Q @ K + bias` has been added. The kernel currently does not support the bias for `softmax(Q @ K) @ V + bias`. We may find a solution in the future if needed.

File: python/tvm/contrib/cutlass/attention_operation.py (hunk 7 above)
Comment (cyx-6) [reply python/tvm/contrib/cutlass/attention_operation.py]: Thanks for your valuable advice! I have set the `false` for these two fields by default. :)

**Review by vinx13 — APPROVED** [review]

File: python/tvm/relax/op/nn/nn.py [hunk 10]
```diff
@@ -577,3 +577,41 @@ def cross_entropy_with_logits(predictions: Expr, labels: Expr) -> Expr:
...
+    key: relax.Expr
+        The input key to the operator. The layout of the input key should be
+        (batch_size, seq_len_kv, num_head, head_dim).
+
+    value: relax.Expr
+        The input value to the operator. The layout of the input value should be
+        (batch_size, seq_len_kv, num_head, head_dim_v).
+
+    bias: Optional[Expr]
+        The optional attention bias to the operator. The layout of the attention bias should be
```
Comment (vinx13) [inline python/tvm/relax/op/nn/nn.py:608]: it's more common to have BNSS' and it allows broadcastable shape. Let's say
```
The shape of the attention bias should be able to broadcast to (batch_size, num_head, seq_len, seq_len_kv)
```

Comment (cyx-6) [reply python/tvm/relax/op/nn/nn.py:608]: Sure! I have updated the layout of the bias to `BNSS'`.

File: python/tvm/contrib/cutlass/attention_operation.py [hunk 11]
```diff
@@ -24,12 +24,12 @@ def instantiate_attention_template(attrs, func_args):
...
     bias_template = """
-  CHECK(${arg3}->ndim == 4); // B, S, N, S'
+  CHECK(${arg3}->ndim == 4); // B, N, S, S'
 
   p.attn_bias_ptr = reinterpret_cast<T *>(${arg3}->data);
-  p.bias_strideH = p.num_keys; // S'
-  p.bias_strideM = p.bias_strideH * p.num_heads; // S' * N
-  p.bias_strideB = p.bias_strideM * p.num_queries; // S' * N * S
+  p.bias_strideM = p.num_keys; // S'
+  p.bias_strideH = p.bias_strideM * p.num_queries; // S' * S
```
Comment (vinx13) [inline python/tvm/contrib/cutlass/attention_operation.py:31]: we should also handle broadcast case

Comment (cyx-6) [reply python/tvm/contrib/cutlass/attention_operation.py:48]: The broadcast for bias is added. Now we accept the bias with layout of `BS'`, `BSS'` and `BNSS'`.

**Review by vinx13 — APPROVED** [review]

**Review by tqchen — APPROVED** [review]

**Comment by masahi** [comment]
The test was added only for cutlass BYOC. Shouldn't we have a standalone test for Relax op + legalize?

**Comment by cyx-6** [comment]
@masahi Yes, as long as it is a Relax op, we should have it fully tested. I will add those necessary tests in the following PR.
